kyuubi-server/src/main/scala/org/apache/kyuubi/credentials/HadoopCredentialsManager.scala

```diff
@@ -0,0 +1,225 @@
+
+  private var engineRefCache: LoadingCache[String, EngineCredentialsRef] = _
+
+  private var providers: Map[String, HadoopDelegationTokenProvider] = _
+  private var requiredProviders: Map[String, HadoopDelegationTokenProvider] = _
+
+  private var renewalInterval: Long = _
+  private var renewalRetryWait: Long = _
+  private var renewalExecutor: ScheduledExecutorService = _
+  private var hadoopConf: Configuration = _
+
+  override def initialize(conf: KyuubiConf): Unit = {
+    val cacheLoader = new CacheLoader[String, EngineCredentialsRef] {
+      override def load(key: String): EngineCredentialsRef = new EngineCredentialsRef
+    }
+    engineRefCache = CacheBuilder.newBuilder()
+      .concurrencyLevel(conf.get(KyuubiConf.SERVER_EXEC_POOL_SIZE))
+      .expireAfterAccess(conf.get(KyuubiConf.ENGINE_IDLE_TIMEOUT), TimeUnit.MILLISECONDS)
+      .build(cacheLoader)
+
+    hadoopConf = KyuubiHadoopUtils.newHadoopConf(conf)
+    providers = loadProviders(conf)
+    requiredProviders = providers.filter { case (_, provider) =>
+      val required = provider.delegationTokensRequired(hadoopConf, conf)
+      if (!required) {
+        info(s"Service ${provider.serviceName} does not require a token." +
+          s" Check your configuration to see if security is disabled or not.")
+      }
+      required
+    }
+    info("Using the following builtin delegation token providers: " +
+      s"${providers.keys.mkString(", ")}.")
+
+    renewalInterval = conf.get(CREDENTIALS_RENEWAL_INTERVAL)
+    renewalRetryWait = conf.get(CREDENTIALS_RENEWAL_RETRY_WAIT)
+    super.initialize(conf)
+  }
+
+  override def start(): Unit = {
+    renewalExecutor =
+      ThreadUtils.newDaemonSingleThreadScheduledExecutor("Delegation Token Renewal Thread")
+    super.start()
+  }
+
+  override def stop(): Unit = {
+    if (renewalExecutor != null) {
+      renewalExecutor.shutdownNow()
+      try {
+        renewalExecutor.awaitTermination(10, TimeUnit.SECONDS)
+      } catch {
+        case _: InterruptedException =>
+      }
+    }
+    super.stop()
+  }
+
+  /**
+   * Send credentials to SQL engine if [[HadoopCredentialsManager]] had a newer version.
+   * `appUser`'s credentials are created if not exist.
+   * [[HadoopCredentialsManager]] takes care created credentials' renewal.
+   * @param engineId SQL engine identifier
+   * @param appUser  User identity that the SQL engine uses.
+   * @param send     Function to send encoded credentials to SQL engine
+   * @return `true` if credentials are sent successfully. Else, false.
+   * @throws KyuubiException if sending credentials fails.
+   */
+  def sendCredentialsIfNeeded(
+      engineId: String,
+      appUser: String,
+      send: String => Unit): Boolean = {
+    require(renewalExecutor != null, "renewalExecutor should be initialized")
+
+    val userRef = getUserRef(appUser)
+    val engineRef = getEngineRef(engineId)
+
+    var sent = false
```

Comment:
where is it used?

Reply:
Used by test code to verify sending result only. 

Removed since sending result already verified by comparing `EngineCredentialsRef#getEpoch` with `UserCredentialsRef#getEpoch`.